internal/internal_worker_usage_collector.go

```diff
@@ -0,0 +1,126 @@
+package internal
+
+import (
+	"context"
+	"github.com/shirou/gopsutil/cpu"
+	"github.com/uber-go/tally"
+	"go.uber.org/cadence/internal/common/metrics"
+	"go.uber.org/zap"
+	"runtime"
+	"sync"
+	"time"
+)
+
+type (
+	workerUsageCollector struct {
+		workerType   string
+		cooldownTime time.Duration
+		logger       *zap.Logger
+		ctx          context.Context
+		wg           *sync.WaitGroup // graceful stop
+		cancel       context.CancelFunc
+		metricsScope tally.Scope
+	}
+
+	workerUsageCollectorOptions struct {
+		Enabled      bool
+		Cooldown     time.Duration
+		MetricsScope tally.Scope
+		WorkerType   string
+	}
+
+	hardwareUsage struct {
+		NumCPUCores     int
+		CPUPercent      float64
+		NumGoRoutines   int
+		TotalMemory     float64
+		MemoryUsedHeap  float64
+		MemoryUsedStack float64
+	}
+)
+
+func newWorkerUsageCollector(
+	options workerUsageCollectorOptions,
+	logger *zap.Logger,
+) *workerUsageCollector {
+	if !options.Enabled {
+		return nil
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	return &workerUsageCollector{
+		workerType:   options.WorkerType,
+		cooldownTime: options.Cooldown,
+		metricsScope: options.MetricsScope,
+		logger:       logger,
+		ctx:          ctx,
+		cancel:       cancel,
+		wg:           &sync.WaitGroup{},
+	}
+}
+
+func (w *workerUsageCollector) Start() {
+	w.wg.Add(1)
+	go func() {
```

Comment:
Do we need to spawn a goroutine per worker? Why not ensure only 1 running?

Reply:
Only the hardware emitting is once per host, all other metrics will be worker-specific. (e.g activity poll response vs. decision poll response)

Reply:
For now I see only w.collectHardwareUsage() which will just spawn bunch of data into the same scope. I would suggest separating hardware emitter and worker specific metrics.

Reply:
That's current design, for each type of metrics based on their origin, I will create a separate gorountine for each of them. But they would be contained under a single workerusagecollector so that their result can be collected and sent in one place